File: systemc/TLM_multisock.cpp
#include "multi-socket-example/initiator.h"
#include "multi-socket-example/bus.h"
#include "multi-socket-example/target.h"

// *****************************************************************************************
// Top-level module instantiates 4 initiators, a bus, and 4 targets
// *****************************************************************************************
const int n = 4;

SC_MODULE(Top)
{
  Initiator* init[n];
  Bus*       bus;
  Target*    target[n];

  SC_CTOR(Top)
  {
    bus   = new Bus("bus");

    // ***************************************************************************
    // bus->init_socket and bus->targ_socket are multi-sockets, each bound 4 times
    // ***************************************************************************

    for (int i = 0; i < n; i++)
    {
      char txt[20];
      sprintf(txt, "init_%d", i);
      init[i] = new Initiator(txt);
      init[i]->socket.bind( bus->targ_socket );
    }

    for (int i = 0; i < n; i++)
    {
      char txt[20];
      sprintf(txt, "target_%d", i);
      target[i] = new Target(txt);

      bus->init_socket.bind( target[i]->socket );
    }
  }
};

int sc_main(int argc, char* argv[])
{
  Top top("top");
  srand(120);
  sc_start();

  cout << "\n***** Messages have been written to file output.txt                    *****\n";
  cout << "***** Select 'Download files after run' to read file in EDA Playground *****\n\n";

  return 0;
}

File: simpy/TLM_multisock.py
from simpy import Environment
from multi_socket_example.initiator import *
from multi_socket_example.bus import *
from multi_socket_example.target import *

class Top(Module):
    def __init__(self, env: Environment, name: str):
        super().__init__(env, name)
        self.bus = Bus(self.env, 'bus')
        self.init = []
        self.target = []
        for i in range(4):
            init = Initiator(self.env, 'init_{}'.format(i))
            init.socket.bind(self.bus.targ_socket)
            self.init.append(init)

            targ = Target(self.env, 'target_{}'.format(i))
            self.bus.init_socket.bind(targ.socket)
            self.target.append(targ)

            dprint(init, init.name)
            dprint(targ, targ.name)


if __name__ == '__main__':
    env = simpy.Environment()
    top = Top(env, 'top')
    env.run()

    print("\n***** Messages have been written to file output.txt                    *****")
